python/ray/_private/test_utils.py

```diff
@@ -1388,136 +1388,240 @@ def teardown_tls(key_filepath, cert_filepath, temp_dir):
+    abstract method that should kill the resource and add it to the `killed` set.
+    """
 
-        def ready(self):
-            pass
+    def __init__(
+        self,
+        head_node_id,
+        kill_interval_s: float = 60,
+        max_to_kill: int = 2,
+        task_filter: Optional[Callable] = None,
+    ):
+        self.kill_interval_s = kill_interval_s
+        self.is_running = False
+        self.head_node_id = head_node_id
+        self.killed = set()
+        self.done = ray._private.utils.get_or_create_event_loop().create_future()
+        self.max_to_kill = max_to_kill
+        self.task_filter = task_filter
+        # -- logger. --
+        logging.basicConfig(level=logging.INFO)
+
+    def ready(self):
+        pass
+
+    async def run(self):
+        self.is_running = True
+        while self.is_running:
+            to_kill = await self._find_resource_to_kill()
+
+            if not self.is_running:
+                break
 
-        async def run(self):
-            self.is_running = True
-            while self.is_running:
-                node_to_kill_ip = None
-                node_to_kill_port = None
-                while node_to_kill_port is None and self.is_running:
-                    nodes = ray.nodes()
-                    alive_nodes = self._get_alive_nodes(nodes)
-                    for node in nodes:
-                        node_id = node["NodeID"]
-                        # make sure at least 1 worker node is alive.
-                        if (
-                            node["Alive"]
-                            and node_id != self.head_node_id
-                            and node_id not in self.killed_nodes
-                            and alive_nodes > 2
-                        ):
-                            node_to_kill_ip = node["NodeManagerAddress"]
-                            node_to_kill_port = node["NodeManagerPort"]
-                            break
-                    # Give the cluster some time to start.
-                    await asyncio.sleep(0.1)
-
-                if not self.is_running:
+            sleep_interval = random.random() * self.kill_interval_s
+            time.sleep(sleep_interval)
+
+            self._kill_resource(*to_kill)
+            if len(self.killed) >= self.max_to_kill:
+                break
+            await asyncio.sleep(self.kill_interval_s - sleep_interval)
+
+        self.done.set_result(True)
+
+    async def _find_resource_to_kill(self):
+        raise NotImplementedError
+
+    def _kill_resource(self, *args):
+        raise NotImplementedError
+
+    async def stop_run(self):
+        was_running = self.is_running
+        self.is_running = False
+        return was_running
+
+    async def get_total_killed(self):
+        """Get the total number of killed resources"""
+        await self.done
+        return self.killed
+
+
+@ray.remote(num_cpus=0)
+class NodeKillerActor(ResourceKillerActor):
```

Comment:
Yeah i've tested it with `multi_node_train_benchmark.py`. From the logs [here](https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_qptpdsrgy28pnv1c61tgvwkmft) it looked like 7 workers were killed. From the dashboard logs it looks like 10 workers were killed (expected, ran with `--max-to-kill 10`). I think it's probably because workers can end successfully before we get to them. The task that we're running in that benchmark (`crop_and_flip_image`) doesn't take very long.

But I was also seeing this in the logs:
```
Failed to unpickle serialized exception7Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/exceptions.py", line 46, in from_ray_exception
    return pickle.loads(ray_exception.serialized_exception)
TypeError: __init__() missing 1 required positional argument: 'pid'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 404, in deserialize_objects
    obj = self._deserialize_object(data, metadata, object_ref)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 293, in _deserialize_object
    return RayError.from_bytes(obj)
```
6 times (probably once for every successful kill?, maybe the last one wasn't logged since the task ended).

Not sure if this is the expected and the result of killing a running worker?

Reply:
@rkooo567 do you know if there's an easy way where we can assert that at least some number of failures happened after the test finishes?

Reply:
@Zandew shall we do this for the new Data test by reading the state API for failed tasks?

Actually for the node failure tests, I think we may be able to get this info from ray.cluster_resources(). It'd be great if you could check that in this PR as well, but if not you can file a follow-up issue with `core` label.

Reply:
I think it would be nice to keep setup_chaos separate from the benchmark scripts so that when we want to test chaos for any benchmark all we need to do is call the setup script beforehand. We could set a timer for the killer and ensure that it kills at least a certain number before the timer runs out, and afterwards assert on the number of actual failures like you described above?

Reply:
Discussed offline that we were not going to test this for benchmarks, a unit test is enough